### PR TITLE
Update URL tests

### PR DESCRIPTION
**Pull request by domenic**
The current URL tests were not up to date with the latest URL spec, especially with regard to file URLs. In addition, they did not test serialization against predetermined results, but instead implemented an incorrect version of the URL Standard's algorithm (this is #2038). Finally, they left a lot of things implicit.

This moves to a new test data format that is explicit for every component of the URL, for maximum coverage. This includes previously-untested components like origin, username, and password.

R=@annevk I assume

### CODE REVIEW

**Comment by tobie** [comment]
Reviewers for this pull request are: @w3c/url-reviewers.

File: url/a-element.html [hunk 1]
```diff
@@ -9,11 +9,11 @@
 var setup = async_test("Loading data…")
 setup.step(function() {
   var request = new XMLHttpRequest()
-  request.open("GET", "urltestdata.txt")
+  request.open("GET", "urltestdata.json")
   request.send()
   request.responseType = "text"
```
Comment (zcorpan) [inline url/a-element.html]: Use `responseType = "json"` instead?

File: url/urltestdata.txt [hunk 2]
```diff
@@ -1,402 +0,0 @@
-# FORMAT NOT DOCUMENTED YET (parser is urltestparser.js)
-
-# Based on http://trac.webkit.org/browser/trunk/LayoutTests/fast/url/script-tests/segments.js
```
Comment (zcorpan) [inline url/urltestdata.txt]: Can we retain the comments somehow? (e.g. as strings between tests)

File: url/a-element.html (hunk 1 above)
Comment (domenic) [inline url/a-element.html]: I wasn't sure it was a good ideal to couple to json responseType support. But I guess all modern browsers have that.

File: url/urltestdata.txt (hunk 2 above)
Comment (annevk) [inline url/urltestdata.txt]: That or maybe in a README. They have been changed a bit too so the reference is no longer that valuable.

Comment (domenic) [inline url/urltestdata.txt]: For now I've restored the comments, except for the ones that were obviously wrong (saying "this should fail" when due to previous updates they no longer fail). Some of them are reasonably valuable, talking about why something fails.

File: url/a-element.html (hunk 1 above)
Comment (domenic) [inline url/a-element.html]: Fixed

**Comment by zcorpan** [comment]
I haven't really reviewed the changes to the tests themselves, but otherwise LGTM. (I would have been easier to review changes to the tests if it was converted to json in one commit and then fixes in another.)

**Comment by domenic** [comment]
@zcorpan @annevk I separated this out into three commits which should make it easier to review.

- The first simply updates to JSON format with no change in test data.
- The second updates the tests for the latest URL Standard, at least according to our reference implementation. This is probably where the closest review is needed.
- The third adds extra tests and tests for the origin, username, password, and host components. It also restores the comments because it was easiest for me to do that at the end instead of trying to preserve them through the diffs.

**Comment by zcorpan** [comment]
> The second updates the tests for the latest URL Standard, at least according to our reference implementation. This is probably where the closest review is needed.

Reviewed, LGTM % @annevk's comment.

**Comment by domenic** [comment]
Updated to mark that one back to failure; filed https://bugs.chromium.org/p/v8/issues/detail?id=4654 to fix Chrome.

**Pull request merged**